### PR TITLE
fix(repl): improve package.json support

### LINKED ISSUE #17929: Deno 1.31.0 REPL errors when launched in a Node project
Setup:

```shell
$ ls
package.json

$ cat package.json
{
  "dependencies": {
    "typescript": "^4.9.5"
  }
}

$ deno
Could not set npm package requirements. Error getting response at https://registry.npmjs.org/typescript for package "typescript": An npm specifier not found in cache: "typescript", --cached-only is specified.
Deno 1.31.0
exit using ctrl+d, ctrl+c, or close()
REPL is running with all permissions allowed.
To specify permissions, run `deno repl` with allow flags.
> 

# Also happens with the new REPL
$ deno repl
Could not set npm package requirements. Error getting response at https://registry.npmjs.org/typescript for package "typescript": An npm specifier not found in cache: "typescript", --cached-only is specified.
Deno 1.31.0
exit using ctrl+d, ctrl+c, or close()
```

This doesn't happen if the `package.json` has no dependencies.

Related: #17928 

After running any script like mentioned in #17928, Deno downloads typescript, and then REPL no longer errors.

### LINKED ISSUE #18494: Disable package.json in deno.json
To put it simply, it is problematic to maintain projects with both native Deno and native Node support, when Deno is constantly referencing `package.json`, which is entirely unnecessary.

I suggest adding an option in `deno.json` to disable `package.json`, or even better, any Node compatibility, like:
```json
{
	"disableNode": true
}
```

Or maybe disable `package.json` or Node support when `deno.json` is discovered, and only enabled when an option is present in `deno.json`, like:
```json
{
	"enableNode": true
}
```

Related:
* #17916
* #17930

### PR DESCRIPTION
**Pull request by dsherret**
1. Fixes a cosmetic issue in the repl where it would display lsp warning messages.
2. Lazily loads dependencies from the package.json on use.
3. Supports using bare specifiers from package.json in the REPL.

Closes #17929
Closes #18494

### CODE REVIEW

**Comment by ry** [comment]
Thanks for the quick fix!

Does this mean that REPL will still look for package.json even before there's an import? Why does it need to do that?

**Comment by dsherret** [comment]
> Does this mean that REPL will still look for package.json even before there's an import? Why does it need to do that?

Yes, it will still do that. It's because the repl internally uses the lsp and the lsp eagerly loads it. The lsp eagerly loads it in order to get npm resolution to be consistent/deterministic regardless of what files are currently open or loaded.

**Comment by dsherret** [comment]
Thinking about it more, we could maybe only use the lockfile as a source of truth for deterministic resolution. I will look into changing this.

File: cli/tools/repl/session.rs [hunk 1]
```diff
@@ -106,7 +106,7 @@ pub fn result_to_evaluation_output(
   match r {
     Ok(value) => value,
     Err(err) => {
-      EvaluationOutput::Error(format!("{} {}", colors::red("error:"), err))
+      EvaluationOutput::Error(format!("{} {:#}", colors::red("error:"), err))
```
Comment (dsherret) [inline cli/tools/repl/session.rs:109]: Outputs the error with context.

**Review by bartlomieju — APPROVED** [review]
LGTM, good fix!

**Pull request merged**